src/webextensions.c

```diff
@@ -0,0 +1,640 @@
+/*
+ * Copyright © 2022 Canonical Ltd
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "config.h"
+
+#include <stdint.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <glib/gi18n.h>
+#include <gio/gunixfdlist.h>
+#include <json-glib/json-glib.h>
+
+#include "session.h"
+#include "webextensions.h"
+#include "request.h"
+#include "permissions.h"
+#include "xdp-dbus.h"
+#include "xdp-impl-dbus.h"
+#include "xdp-utils.h"
+
+#define PERMISSION_TABLE "webextensions"
+
+typedef struct _WebExtensions WebExtensions;
+typedef struct _WebExtensionsClass WebExtensionsClass;
+
+struct _WebExtensions
+{
+  XdpDbusWebExtensionsSkeleton parent_instance;
+};
+
+struct _WebExtensionsClass
+{
+  XdpDbusWebExtensionsSkeletonClass parent_class;
+};
+
+static XdpDbusImplAccess *access_impl;
+static WebExtensions *web_extensions;
+
+GType web_extensions_get_type (void);
+static void web_extensions_iface_init (XdpDbusWebExtensionsIface *iface);
+
+G_DEFINE_TYPE_WITH_CODE (WebExtensions, web_extensions, XDP_DBUS_TYPE_WEB_EXTENSIONS_SKELETON,
+                         G_IMPLEMENT_INTERFACE (XDP_DBUS_TYPE_WEB_EXTENSIONS,
+                                                web_extensions_iface_init));
+
+typedef enum _WebExtensionsSessionState
+{
+  WEB_EXTENSIONS_SESSION_STATE_INIT,
+  WEB_EXTENSIONS_SESSION_STATE_STARTING,
+  WEB_EXTENSIONS_SESSION_STATE_STARTED,
+  WEB_EXTENSIONS_SESSION_STATE_CLOSED,
+} WebExtensionsSessionState;
+
+typedef struct _WebExtensionsSession
+{
+  Session parent;
+
+  WebExtensionsSessionState state;
+
+  GPid child_pid;
+  guint child_watch_id;
+
+  int standard_input;
+  int standard_output;
+  int standard_error;
+} WebExtensionsSession;
+
+typedef struct _WebExtensionsSessionClass
+{
+  SessionClass parent_class;
+} WebExtensionsSessionClass;
+
+GType web_extensions_session_get_type (void);
+
+G_DEFINE_TYPE (WebExtensionsSession, web_extensions_session, session_get_type ());
+
+static void
+web_extensions_session_init (WebExtensionsSession *session)
+{
+  session->child_pid = -1;
+  session->child_watch_id = 0;
+
+  session->standard_input = -1;
+  session->standard_output = -1;
```

Comment:
I think that `session->standard_error` needs to be initalized here too.

Reply:
That was it, very nice catch @brianfcoleman !
I rebuilt the portal with this change, and I'm no longer seeing the initial problem where the portal doesn't function correctly until it is killed and respawned.